pkg: add SPIFFS support

vincent-d opened the pull request:
This PR adds support for SPIFFS as a package based on @gebart recent VFS work.

This is still WIP and all functions are not implemented right now.
Some unit tests may be used to test it.

This is based on ~~#5616~~ and uses ~~#5624~~ for the low level.

gebart commented:
The patch for the const buffer in spiffs_write should be upstreamed, if possible. I don't see a reason for why a source buffer in a write situation should be writable.

vincent-d commented:
@gebart I agree, I'll open a PR on the upstream project, but I wanted to have something working.

vincent-d commented:
opendir/readdir/closedir support added. No support for mkdir/rmdir as SPIFFS does not support directory, but has a flat structure and I'm not 100% sure about the behavior of readdir.

Still TODO: find a better way to use spiffs_config.h. Some part of this file should be set in a common part instead of in the board.

gebart commented:
>Still TODO: find a better way to use spiffs_config.h. Some part of this file should be set in a common part instead of in the board.

How about since all defines in spiffs_config.h are already surrounded by `#ifndef`, we add `#include "spiffs_board_config.h"` to the top and let the board specifics come first, then fall back to the defaults?

gebart commented:
just realized that the above idea will cause problems with missing files. We could just add `#include "board.h"` at the top instead

gebart commented:
for reference: https://github.com/pellepl/spiffs/pull/94

vincent-d commented:
Added rename support and rebased on top of #5616

vincent-d commented:
https://github.com/OTAkeys/RIOT/pull/1 merged and rebased on top of #5616

gebart commented on tests/unittests/tests-spiffs/tests-spiffs.c:
```diff
@@ -0,0 +1,309 @@
+    vfs_dirent_t entry;
+    res = vfs_readdir(&dirp, &entry);
+    TEST_ASSERT(res == 1);
+    TEST_ASSERT_EQUAL_STRING("/test0.txt", &(entry.d_name[0]));
```
The order of these files may not be the same order as the order in which they were created. It is also possible that there are other files left over from previous runs or failed tests in the same run which interfere with this check. The test should only verify that all the created files are reported back by `readdir`, the order should be irrelevant.

gebart commented on pkg/spiffs/fs/spiffs_fs.c:
```diff
@@ -0,0 +1,429 @@
+    struct spiffs_dirent e;
+    struct spiffs_dirent *ret;
+
+    memcpy(&d, dirp->private_data.buffer, sizeof(d));
```
What is the purpose of the memcpy back and forth for the call to `SPIFFS_readdir`?

gebart commented on tests/unittests/tests-spiffs/tests-spiffs.c:
```diff
@@ -0,0 +1,309 @@
+
+static void tests_spiffs_write2(void)
+{
+    char buf[1024];
```
This overflows the stack of the unit tests on anything but native.

gebart replied:
adding `static` fixes it (but it's still kinda big)

gebart commented:
There's a problem with the unit test: Running it after a failed run may have left files around in the SPIFFS file system, so there should be some clean up at the beginning of each test to ensure that no leftover files are interfering with the outcome.

vincent-d commented:
Unit tests improved.
Rebased with last mtd commits.

vincent-d commented:
rebased on top of #5616

vincent-d commented:
Rebased on top of #5624

miri64 commented:
Postponed due to Feature Freeze.

vincent-d commented:
Rebased on top of #5624

vincent-d commented:
rebased.

I think it's ready for review.

OlegHahm commented:
Should I remove the WIP label?

vincent-d commented:
@OlegHahm Yes ;)

kYc0o commented:
@vincent-d you can also squash this one if you have time, as far as I can see there's no review yet.

vincent-d commented:
SPIFFS part is squashed

gebart commented:
Needs rebase

vincent-d commented:
I can't find how to solve the build issues.

I had to add `CFLAGS += -std=c11` in the pkg Makefile, to avoid having an error with unamed structs. But this option is not available on msp430 gcc.

Does anyone know how to solve that ? I'm not even sure we will be able to compile spiffs code for msp430 anyway.

gebart commented:
Did you try -std=gnu99?

vincent-d commented:
Just tried it and I have the same issue:
`tests/unittests/bin/pkg/native/spiffs/riotbuild/spiffs_nucleus.h:388:6: error: ISO C99 doesn’t support unnamed structs/unions [-Werror=pedantic]`

vincent-d commented:
Rebased on master since #5624 has been merged

gebart commented on pkg/spiffs/Makefile:
```diff
@@ -0,0 +1,19 @@
+PKG_NAME=spiffs
+PKG_URL=https://github.com/pellepl/spiffs.git
+PKG_VERSION=39937743fbbec4b82308ee08332bf9180408d23b
```
Are there any new updates upstream which we should add?

vincent-d replied:
This is the last release (v0.3.6) I'm not sure we should use a random commit from the master branch. I'd say we stick to this one.

gebart replied:
sorry, I did not see the tag on the Github mobile page. I agree this is the commit we should use.

vincent-d commented on pkg/spiffs/Makefile:
```diff
@@ -0,0 +1,19 @@
+PKG_NAME=spiffs
+PKG_URL=https://github.com/pellepl/spiffs.git
+PKG_VERSION=39937743fbbec4b82308ee08332bf9180408d23b
+PKG_BUILDDIR ?= $(BINDIRBASE)/pkg/$(BOARD)/$(PKG_NAME)
```
Addressed

gebart commented on tests/unittests/tests-spiffs/tests-spiffs.h:
```diff
@@ -0,0 +1,37 @@
+
+#ifdef __cplusplus
+extern "C" {
+    #endif
```
wrong indentation

gebart commented on tests/unittests/tests-spiffs/tests-spiffs.h:
```diff
@@ -0,0 +1,37 @@
+     */
+    void tests_spiffs(void);
+
+    #ifdef __cplusplus
```
wrong indentation

vincent-d commented:
Fixed indentation issues in tests.

gebart commented on pkg/spiffs/fs/spiffs_fs.c:
```diff
@@ -0,0 +1,486 @@
+    if (flags & O_WRONLY) {
+        s_flags |= SPIFFS_O_WRONLY;
+    }
+    if (flags & O_RDWR) {
```
Is it possible that O_RDWR could be implemented as a bitmask combination of read and write? (like how SPIFFS implements SPIFFS_O_RDWR)
if so, we need to make sure that it matches exactly:

```c
if ((flags & O_RDWR) == O_RDWR) {
```

gebart replied:
The other flags should probably be updated accordingly

gebart replied:
There's also a SPIFFS_O_RDONLY which should be set 
```c
if ((flags & O_RDONLY) == O_RDONLY) {
    s_flags |= SPIFFS_O_RDONLY;
}
```
(must be checked for equality because O_RDONLY is defined as 0 by default in newlib)

vincent-d commented:
Unittests are failing when using the dummy mtd. I removed the wrong `#undef MTD_0` and tests are running successfully on native. On mulle you could define `MTD_0` to point to the mtd device.

gebart commented:
@vincent-d right, I will post a few fixes and updates for Mulle in the mtd spi nor PR (#6762)

vincent-d commented:
@gebart flags translation fixed, nice spotted ;)

gebart commented on pkg/spiffs/fs/spiffs_fs.c:
```diff
@@ -222,28 +222,31 @@ static int _open(vfs_file_t *filp, const char *name, int flags, mode_t mode, con
-    spiffs_flags s_flags = SPIFFS_O_RDONLY;
-    if (flags & O_APPEND) {
+    spiffs_flags s_flags = 0;
+    if ((flags & O_RDONLY) == O_RDONLY) {
```
Sorry to comment once again on what I just told you to change. I realize that we will set SPIFFS_O_RDONLY on files even when they are opened with O_WRONLY.
We should change this to use O_ACCMODE as mask instead. O_ACCMODE is defined as the combined mask (O_WRONLY|O_RDONLY|O_RDWR)

vincent-d replied:
Addressed

vincent-d commented:
Fixed unittests:
 - Disable `tests-spiffs` for msp430 boards (btw, fixed the `DISABLE_TEST_FOR_MSP430` usage)
 - Add some boards to `BOARD_INSUFFICIENT_MEMORY`

vincent-d commented:
I don't understand why Murdock2 fails building unittests with that error:
`"make" -C /tmp/dwq.0.6264584041516473/69df5fe7980473b740b005c43249a21c/build/pkg/spiffs/riotbuild
In file included from /tmp/dwq.0.6264584041516473/69df5fe7980473b740b005c43249a21c/pkg/spiffs/fs/spiffs_fs.c:26:0:
/tmp/dwq.0.6264584041516473/69df5fe7980473b740b005c43249a21c/sys/include/fs/spiffs_fs.h:27:20: fatal error: spiffs.h: No such file or directory
 #include "spiffs.h"
                    ^
compilation terminated.`

while Murdock and Jenkins are fine and I can't reproduce that locally either.

gebart commented:
Could it be a race in the git checkout in Murdock2?

vincent-d commented:
Rebased, will be easier to test with NOR flashes

vincent-d commented:
@kaspar030 do you have a clue why Murdock2 is failling ?

kaspar030 commented:
> @kaspar030 do you have a clue why Murdock2 is failling ?

You need to fix the include path in ```pkg/spiffs/Makefile.include```:

replace

    INCLUDES += -I$(BINDIRBASE)/pkg/$(BOARD)/spiffs/riotbuild

with

    INCLUDES += -I$(PKGDIRBASE)/spiffs/riotbuild

vincent-d commented:
Done, thanks @kaspar030

vincent-d commented:
Unittests seem OK now

gebart commented:
The unit tests take ages to run on Mulle with a real SPI NOR flash memory. Will do some more investigation.

gebart commented:
@vincent-d what's the run time length on your platform?
Mulle took several minutes to run the entire spiffs unit tests, most of it was spent in the write2 test.

vincent-d commented:
@gebart the write2 test was initially there to ensure erase works correctly, so we need to fill the whole memory before the first erase actually happen, that's why it takes ages... Maybe we should just remove that test

gebart commented:
@vincent-d I think it makes sense as a test, but it slows down the execution of the unit tests too much. Maybe only enable it on native for now? 

I think it would be great if we had a separate benchmark test for spiffs, to compare read/write speeds between commits and between platforms. The write2 test could be a part of that kind of application as well.

gebart commented on tests/unittests/tests-spiffs/tests-spiffs.c:
```diff
@@ -0,0 +1,413 @@
+
+static void test_spiffs_teardown(void)
+{
+    vfs_mount(&_test_spiffs_mount);
```
There's a lot of mounting and unmounting going on, what do you think about mounting in the setup phase, saving the result to a file-global (i.e. static global) variable and just checking the variable inside the mount_unmount test?

gebart replied:
The unmount would need to be tested inside the mount_unmount test though

vincent-d replied:
W could mount in the `test_spiffs_setup`, unmount in the `test_spiffs_teardown` and change the mount_unmount test to a unmount_mount. What do you think ?

gebart replied:
Sure, that would work I guess. I forgot, do we handle unmounting an already unmounted fs in VFS?

vincent-d replied:
I guess either the fs implementation or removing the mountpoint from the list will fail, but it will fail silently.

gebart commented on tests/unittests/tests-spiffs/tests-spiffs.c:
```diff
@@ -0,0 +1,413 @@
+    vfs_unlink("/test-spiffs/test0.txt");
+    vfs_unlink("/test-spiffs/test1.txt");
+    vfs_unlink("/test-spiffs/a/test2.txt");
+    vfs_umount(&_test_spiffs_mount);
```
... and a flag variable to tell whether we need to unmount or not.

gebart commented on tests/unittests/tests-spiffs/tests-spiffs.c:
```diff
@@ -0,0 +1,413 @@
+    TEST_ASSERT(mp >= 0);
+
+    int res;
+    for (int j = 0; j < 5; j++) {
```
This needs a comment explaining what we want to test.
I don't really see how the magic numbers in this test ensure that we fill the memory and need to erase sectors..

gebart commented on tests/unittests/tests-spiffs/tests-spiffs.c:
```diff
@@ -0,0 +1,413 @@
+            TEST_ASSERT_EQUAL_INT(sizeof(buf), res);
+        }
+
+        res = vfs_lseek(fd, 0, SEEK_SET);
```
is this line necessary?

vincent-d commented:
For the write2 test, I used the magic numbers with my memory. But I already reduced them to avoid the timeout on native, which does not really make sense, I agree.
I propose we just drop this test for now to not block that PR and we prepare a separate test app which could be used for benchmarking.

vincent-d commented:
Fix unittests
- write2 removed
- comments addressed

vincent-d commented:
Squashed

gebart commented:
Did another test run on Mulle and native, both work fine.

gebart commented:
Murdock is fine, go!

gebart commented:
Thanks, @vincent-d! Finally RIOT has a file system out of the box! 🎉

vincent-d commented:
Great! Thanks @gebart
